subsys/usb/usb_device.c

```diff
@@ -898,12 +922,19 @@ int usb_set_config(struct usb_cfg_data *config)
 					     config->interface.class_handler,
 					     config->interface.payload_data);
 	}
-	/* register vendor request handlers */
-	if (config->interface.vendor_handler) {
+
+	/* register vendor request handler */
+	if (config->interface.vendor_handler || usb_os_desc_enabled()) {
```

Comment:
There is already a wrapper for composite configuration:
https://github.com/zephyrproject-rtos/zephyr/pull/7916/commits/b09e5dcb0e56f5ca6800938fdb27a91dbc194e8b#diff-609952c961f3c77cb7811bedee4bfd2cL1388

That also needs to be changed:
https://github.com/zephyrproject-rtos/zephyr/pull/7916/commits/b09e5dcb0e56f5ca6800938fdb27a91dbc194e8b#diff-609952c961f3c77cb7811bedee4bfd2cR1494

We can simplify it all and only use usb_composite_init for initialization without explicitly turning on CONFIG_USB_COMPOSITE_DEVICE, but first interface descriptors in samples webusb and wpanusb have to be reworked.

Reply:
@finikorg any comments on this? I think this makes sense

Reply:
@jfischer-phytec-iot @carlescufi vendor_handler() in usb_devices.c does not handle request correctly. It expects that setup->wIndex indicates interface number, which is not the case for some (maybe all) of requests. 
"Microsoft OS 2.0 Descriptors Specification":
```
wIndex 0x7 for MS_OS_20_DESCRIPTOR_INDEX.
wIndex 0x8 for MS_OS_20_SET_ALT_ENUMERATION
...
etc
```

Reply:
@jfischer-phytec-iot Added patch #afe05be

Reply:
@jfischer-phytec-iot Please have a look at 96a9583